### ISSUE TITLE
Input on Classic NES games seems abnormal

### ISSUE THREAD

**Issue by zambonin** [mgba-emu/mgba#232]
Hello,

I noticed a problem when testing a few of the Classic NES games on mGBA (specifically, Zelda I, II and Metroid; Super Mario Bros. works with no apparent flaws).

I'm running Arch Linux, and tested with mGBA 0.4.0 from the [official repository](https://www.archlinux.org/packages/community/x86_64/mgba-sdl/) and compiled r3500 from the [user repository](https://aur.archlinux.org/packages/mgba-sdl-git/), to no avail  -- both present the same problem, which is described below, on my machine.

For some reason, either version won't run/compile on another machine also running Arch Linux (due to `libavcodec` version mismatch), so I can't reproduce this on another Linux machine, however 0.4.0 on Windows also has the same input problem.

Any time input is needed, I have to use the fast-forward feature and keep pressing the button. Link or Samus barely move, even if I hold Left/Right + TAB; same with character selection on save creation screens, or the "press START" screen.

If this problem can be reproduced, I can take some time and see if it happens with more Classic NES games. Otherwise, I apologize in advance for the lack of proper testing.

Thanks for your time.

**Comment by endrift** [comment]
Strange, I can't seem to reproduce this.

**Comment by zambonin** [comment]
I briefly compiled and installed 0.3.2 and it seems to be persistent. 

According to your [post](https://mgba.io/2014/12/28/classic-nes/) about the games, they should be working in 0.2.0 - but the problem persists.

On version 0.4.0, a lot of errors are displayed on the logs (0.2.0 said that the gamepak wasn't implemented so I figure that wouldn't be very helpful). I will paste some of those so you can see if it's relevant.

```
GAME ERROR:	Out of bounds ROM Load8: 0x083001A0
GAME ERROR:	Out of bounds ROM Load8: 0x083001A0
GAME ERROR:	Out of bounds ROM Load8: 0x083001A0
GAME ERROR:	Out of bounds ROM Load8: 0x083001A0
GAME ERROR:	Out of bounds ROM Load8: 0x083001A0
GAME ERROR:	Read from write-only I/O register: 0DC
SWI:	SWI: 02 r0: 06006800 r1: 07000000 r2: 07000010 r3: 08201C3D
SWI:	SWI: 02 r0: 06006800 r1: 07000000 r2: 07000010 r3: 08201C3D
SWI:	SWI: 02 r0: 06006800 r1: 07000000 r2: 07000010 r3: 08201C3D
GAME ERROR:	Out of bounds ROM Load8: 0x083001A1
GAME ERROR:	Out of bounds ROM Load8: 0x083001A1
GAME ERROR:	Out of bounds ROM Load8: 0x083001A1
GAME ERROR:	Out of bounds ROM Load8: 0x083001A1
GAME ERROR:	Out of bounds ROM Load8: 0x083001A1
GAME ERROR:	Read from write-only I/O register: 0DC
SWI:	SWI: 02 r0: 06003200 r1: 07000000 r2: 07000010 r3: 08201C3D
SWI:	SWI: 02 r0: 06003200 r1: 07000000 r2: 07000010 r3: 08201C3D
SWI:	SWI: 02 r0: 06003200 r1: 07000000 r2: 07000010 r3: 08201C3D
```
The SW interrupt alternates between `06003200` and `06006800` on `r0`, while the out of bounds increments by one every time.

**Comment by zambonin** [comment]
```
                  ROM                                                   mGBA 0.4.0

Classic NES Series - Bomberman (U) [f1].gba                             game pak error
Classic NES Series - Bomberman (U) [f_4].gba                            wrong input
Classic NES Series - Bomberman (U) [f_5].gba                            wrong input
Classic NES Series - Bomberman (U).gba                                  wrong input
Classic NES Series - Bomberman (U) [o1][f1].gba                         game pak error
Classic NES Series - Bomberman (U) [o1].gba                             works as expected

Classic NES Series - Castlevania (U) [f1].gba                           game pak error
Classic NES Series - Castlevania (U) [f_4].gba                          wrong input
Classic NES Series - Castlevania (U) [f_5].gba                          wrong input
Classic NES Series - Castlevania (U).gba                                wrong input
Classic NES Series - Castlevania (U) [o1][f1].gba                       game pak error
Classic NES Series - Castlevania (U) [o1].gba                           works as expected

Classic NES Series - Donkey Kong (U) [f1].gba                           game pak error
Classic NES Series - Donkey Kong (U) [f_4].gba                          wrong input
Classic NES Series - Donkey Kong (U) [f_5].gba                          wrong input
Classic NES Series - Donkey Kong (U).gba                                wrong input
Classic NES Series - Donkey Kong (U) [o1][f1].gba                       game pak error
Classic NES Series - Donkey Kong (U) [o1][f_4].gba                      works as expected
Classic NES Series - Donkey Kong (U) [o1][f_5].gba                      works as expected
Classic NES Series - Donkey Kong (U) [o1].gba                           works as expected

Classic NES Series - Dr. Mario (U) [f1].gba                             game pak error
Classic NES Series - Dr. Mario (U) [f_4].gba                            wrong input
Classic NES Series - Dr. Mario (U) [f_5].gba                            wrong input
Classic NES Series - Dr. Mario (U).gba                                  wrong input
Classic NES Series - Dr. Mario (U) [o1][f1].gba                         game pak error
Classic NES Series - Dr. Mario (U) [o1].gba                             works as expected

Classic NES Series - Excitebike (U) [f1].gba                            game pak error
Classic NES Series - Excitebike (U) [f_4].gba                           wrong input
Classic NES Series - Excitebike (U) [f_5].gba                           wrong input
Classic NES Series - Excitebike (U).gba                                 wrong input
Classic NES Series - Excitebike (U) [o1][f1].gba                        game pak error
Classic NES Series - Excitebike (U) [o1].gba                            works as expected
*graphical flickering on the blue borders and other elements 

Classic NES Series - Ice Climber (U) [f1].gba                           game pak error
Classic NES Series - Ice Climber (U) [f_4].gba                          wrong input
Classic NES Series - Ice Climber (U) [f_5].gba                          wrong input
Classic NES Series - Ice Climber (U) [!].gba                            wrong input
Classic NES Series - Ice Climber (U) [o1][f1].gba                       game pak error
Classic NES Series - Ice Climber (U) [o1][f_4].gba                      works as expected
Classic NES Series - Ice Climber (U) [o1][f_5].gba                      works as expected
Classic NES Series - Ice Climber (U) [o1].gba                           works as expected
Classic NES Series - Ice Climber (U) [o2].gba                           works as expected

Classic NES Series - Metroid (U) [f1].gba                               game pak error
Classic NES Series - Metroid (U) [f_4].gba                              wrong input
Classic NES Series - Metroid (U) [f_5].gba                              wrong input
Classic NES Series - Metroid (U) [!].gba                                wrong input
Classic NES Series - Metroid (U) [o1][f1].gba                           game pak error
Classic NES Series - Metroid (U) [o1].gba                               works as expected
Classic NES Series - Metroid (U) [o2].gba                               works as expected

Classic NES Series - Pac-Man (U) [f1].gba                               game pak error
Classic NES Series - Pac-Man (U) [f_4].gba                              wrong input
Classic NES Series - Pac-Man (U) [f_5].gba                              wrong input
Classic NES Series - Pac-Man (U).gba                                    wrong input
Classic NES Series - Pac-Man (U) [o1][f_4].gba                          works as expected
Classic NES Series - Pac-Man (U) [o1][f_5].gba                          works as expected
Classic NES Series - Pac-Man (U) [o1].gba                               works as expected

Classic NES Series - Super Mario Bros. (U) [f1].gba                     game pak error
Classic NES Series - Super Mario Bros. (U) [f_4].gba                    wrong input
Classic NES Series - Super Mario Bros. (U) [f_5].gba                    wrong input
Classic NES Series - Super Mario Bros. (U).gba                          wrong input
Classic NES Series - Super Mario Bros. (U) [o1][f1].gba                 game pak error
Classic NES Series - Super Mario Bros. (U) [o1].gba                     works as expected
*graphical flickering easily perceived with fast-forward

Classic NES Series - The Legend of Zelda (U) [b1].gba                   game pak error
Classic NES Series - The Legend of Zelda (U) [f1][a1].gba               seg. fault
Classic NES Series - The Legend of Zelda (U) [f1][a2].gba               game pak error
Classic NES Series - The Legend of Zelda (U) [f1][f_4].gba              game pak error
Classic NES Series - The Legend of Zelda (U) [f1][f_5].gba              game pak error
Classic NES Series - The Legend of Zelda (U) [f1] (Save Type).gba       seg. fault
Classic NES Series - The Legend of Zelda (U) [f2].gba                   wrong input
Classic NES Series - The Legend of Zelda (U) [f3].gba                   game pak error
Classic NES Series - The Legend of Zelda (U) [f_4].gba                  wrong input
Classic NES Series - The Legend of Zelda (U) [f_5].gba                  wrong input
Classic NES Series - The Legend of Zelda (U).gba                        wrong input
Classic NES Series - The Legend of Zelda (U) [o1][f1].gba               works as expected
Classic NES Series - The Legend of Zelda (U) [o1][f2].gba               works as expected
Classic NES Series - The Legend of Zelda (U) [o1].gba                   works as expected
Classic NES Series - The Legend of Zelda (U) [o2][f1].gba               game pak error
("The game has crashed with the following error: Jumped to invalid address 08100000")

Classic NES Series - Xevious (U) [f1].gba                               game pak error
Classic NES Series - Xevious (U) [f_4].gba                              wrong input
Classic NES Series - Xevious (U) [f_5].gba                              wrong input
Classic NES Series - Xevious (U).gba                                    wrong input
Classic NES Series - Xevious (U) [o1][f_4].gba                          works as expected
Classic NES Series - Xevious (U) [o1][f_5].gba                          works as expected
Classic NES Series - Xevious (U) [o1].gba                               works as expected

Classic NES Series - Zelda II - The Adventure of Link (U) [f1].gba      game pak error
Classic NES Series - Zelda II - The Adventure of Link (U) [f_4].gba     wrong input
Classic NES Series - Zelda II - The Adventure of Link (U) [f_5].gba     wrong input
Classic NES Series - Zelda II - The Adventure of Link (U) [!].gba       wrong input
Classic NES Series - Zelda II - The Adventure of Link (U) [o1][f1].gba  game pak error
Classic NES Series - Zelda II - The Adventure of Link (U) [o1].gba      works as expected
Classic NES Series - Zelda II - The Adventure of Link (U) [o2].gba      works as expected
```

As you can see, there's quite a pattern. Maybe your dump is too good? :)

The results seem consistent on r3512, I didn't test all of ROMs again though.

**Comment by endrift** [comment]
Fixed in aa0ee743b3779e056b74b15dba9415c541a6533f. Turns out my dumps were overdumps. Can you please look into the segfault a bit more and file a separate bug if it still reproduces?
